### DISCUSSION TITLE
Added a WaveStream audio decoder

### DISCUSSION THREAD

**Discussion by Mbucari** [Mbucari/AAXClean.Codecs#5] [Announcements]
@audiamus
I was working on a new feature for Libation that would allow streaming audio, but progress has stalled due to the difficulties of cross-platform audio playback.

However, the basic building blocks have been completed are are included in verion 1.0, and I thought you may be interested.

To get a [WaveStream](https://markheath.net/post/naudio-wavestream-in-depth), simply call `Mp4File.GetWaveStream(TimeSpan)`. The TimeSpan parameter is the amount of audio time to buffer. You can use it to play decrypted local m4b files or stream and play Aax files (decrypted on the fly).  I don't know if your apps support downloading directly from audible's servers, but if they do you should know that their servers support Http byte ranges. You can use ranges to make a stream that supports seeking.

**Comment by audiamus** [comment]
>  I don't know if your apps support downloading directly from audible's servers, but if they do you should know that their servers support Http byte ranges.

Interesting. I am currently downloading with the standard network stream, with a rolling copy to file stream, to allow progress report. I initially tried to feed the network stream directly into into your decoder, but that failed, of course, because it's not seekable. However, I don't think I will change my download mechanism.

**Comment by Mbucari** [comment]
> initially tried to feed the network stream directly into into your decoder, but that failed, of course, because it's not seekable.

That's strange. From the beginning I designed aax clean to be able to decode from a network stream without having to write to a file first. It never needs to seek backwards, but does need to seek forward by a maximum of a few hundred bytes. If a stream doesn't support seeking, it just reads and discards those few hundred bytes to get to the next Target location.

**Comment by audiamus** [comment]
It may have been an issue with a particular version back then, when I tested it the first time. I left it untouched once it compiled. The only change I made since was to adapt to the new async interface after you introduced it.

**Comment by Mbucari** [comment]
I think I know what the problem was. If a stream doesn't support seeking, you have to use the (undocumented)
`AaxFile(Stream file, long fileSize, bool additionalFixups = true)` overload. AAXClean needs to know the content length to work, and Stream.Length is unsupported for network streams.

**Reply by audiamus** [reply]
I see. That makes sense. I vaguely remember it took me some time to figure out the download file length, as I needed it to report progress.
